source/vir/src/interpreter.rs

```diff
@@ -0,0 +1,1391 @@
+    fn insert_call(&mut self, f: &Fun, args: &Exps, result: &Exp, memoize: bool) {
+        if self.enable_cache && memoize {
+            self.cache.entry(f.clone()).or_default().insert(args.into(), result.clone());
+        }
+    }
+
+    fn lookup_call(&mut self, f: &Fun, args: &Exps, memoize: bool) -> Option<Exp> {
+        if self.enable_cache && memoize {
+            if self.perf {
+                let count = self.fun_calls.entry(f.clone()).or_default();
+                *count += 1;
+            }
+            self.cache.get(f)?.get(&args.into()).cloned()
+        } else {
+            None
+        }
+    }
+
+    fn log(&self, s: String) {
+        if self.debug {
+            let mut log = self.log.as_ref().unwrap();
+            writeln!(log, "{}", s).expect("I/O error writing to the interpreter's log");
+        }
+    }
+}
+
+/// Static context for the interpreter
+struct Ctx<'a> {
+    /// Maps each function to the SST expression representing its body
+    fun_ssts: &'a SstMap,
+    /// We avoid infinite loops by running for a fixed number of intervals
+    max_iterations: u64,
+    /// Number of bits we assume the underlying architecture supports
+    arch_size_min_bits: u32,
+}
+
+/// Interpreter-internal expressions
+#[derive(Debug, Clone)]
+pub enum InterpExp {
+    /// Track free variables (those not introduced inside an assert_by_compute),
+    /// so they don't get confused with bound variables.
+    FreeVar(UniqueIdent),
+    /// Optimized representation of intermediate sequence results
+    Seq(Vector<Exp>),
+}
+
+/*****************************************************************
+ * Functionality needed to compute equality between expressions  *
+ *****************************************************************/
+
+/// Trait to compute syntactic equality of two objects.
+trait SyntacticEquality {
+    /// Compute syntactic equality. Returns `Some(b)` if syntactically, equality can be guaranteed,
+    /// where `b` iff `self == other`. Otherwise, returns `None`.
+    fn syntactic_eq(&self, other: &Self) -> Option<bool>;
+    fn definitely_eq(&self, other: &Self) -> bool {
+        matches!(self.syntactic_eq(other), Some(true))
+    }
+    fn definitely_ne(&self, other: &Self) -> bool {
+        matches!(self.syntactic_eq(other), Some(false))
+    }
+    /// If we cannot definitively establish equality, we conservatively return `None`
+    fn conservative_eq(&self, other: &Self) -> Option<bool> {
+        self.definitely_eq(other).then(|| true)
+    }
+}
+
+// Automatically get syntactic equality over typs, exps, ... once we know syntactic equality over typ, exp, ...
+impl<T: SyntacticEquality> SyntacticEquality for Arc<Vec<T>> {
+    fn syntactic_eq(&self, other: &Self) -> Option<bool> {
+        if self.len() != other.len() {
+            Some(false)
+        } else {
+            self.iter()
+                .zip(other.iter())
+                .try_fold(true, |acc, (l, r)| Some(acc && l.syntactic_eq(r)?))
```

Comment:
> It looks like this always returns None, but if for any index it returns Some(false), then you can definitively return Some(false), even if some other index is indeterminate.

Ah, you're saying that if any index returns `None`, then the `l.syntactic_eq(r)?` will turn the entire result to `None`?   Looking at the [docs](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold), do I need to use the `ControlFlow` construct instead, if I want to stick with `try_fold`?

As a side note, I suspect it will be rare that we can establish `definitely_ne`, given that we're operating at the syntactic level.

Reply:
I added some additional logic to handle this case.